File: upstatd/sysbat.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

type sysbat struct {
	engNow, engFul string
	fullChar       rune
	thresholds     []float32
	usedChars      []rune
}

// TODO -- make the full char configurable from the poly config
func sysStruct() sysbat {
	sb := sysbat{}
	sb.engNow = "/sys/class/power_supply/BAT0/energy_now"
	sb.engFul = "/sys/class/power_supply/BAT0/energy_full"
	sb.fullChar = '⚡'
	sb.thresholds = []float32{98.9, 88.0, 70.0, 50.0, 24.0}
	sb.usedChars = []rune{' ', '', '', '', '', ''}
	return sb
}

func readFd(f string, buf []byte) (float32, error) {
	fd, err := os.Open(f)
	if err != nil {
		return 0.0, err
	}
	n, err := fd.Read(buf)
	if err != nil {
		return 0.0, err
	}
	s := string(buf[:n-1]) //slice off newline
	eng, err := strconv.ParseFloat(s, 32)
	if err != nil {
		return 0.0, err
	}
	eval := float32(eng)
	return eval, nil
}

func (sb *sysbat) getCharge() string {
	buf := make([]byte, 20)
	enow, err := readFd(sb.engNow, buf)
	if err != nil {
		polyerr(err)
		return ""
	}
	eful, err := readFd(sb.engFul, buf)
	if err != nil {
		polyerr(err)
		return ""
	}
	used := (enow / eful) * 100.0
	str := "BAT|"
	if used > sb.thresholds[0] {
		return fmt.Sprintf("%s%c", str, sb.fullChar)
	}

	for i := 1; i < 5; i++ {
		if used > sb.thresholds[i] {
			return fmt.Sprintf(
				"%s%.2f %c",
				str, used, sb.usedChars[i],
			)
		}
	}

	return fmt.Sprintf(
		"%s%.2f %c",
		str, used, sb.usedChars[5],
	)
}

File: upstatd/src/sysbat.rs
use std::{fs::File, io::Read};

use crate::colors;

static ENG_NOW: &'static str = "/sys/class/power_supply/BAT0/energy_now";
static ENG_FUL: &'static str = "/sys/class/power_supply/BAT0/energy_full";
static LIGHTNING_BOLT: &'static str = "⚡";
static THRESHOLDS: [f32; 5] = [98.9, 88.0, 70.0, 50.0, 24.0];
static UCHARS: [char; 6] = [' ', '', '', '', '', ''];

fn get_val(path: &'static str, buf: &mut [u8; 20]) -> f32 {
    let mut f = File::open(path).unwrap();
    let n = f.read(buf).unwrap();
    let s = String::from_utf8(
        buf[..n - 1].to_vec(), // slice off newline, causes issues with Rust's parser
    )
    .unwrap();
    let val: f32 = s.parse().unwrap();
    return val;
}

pub fn print_bat(cs: &colors::Colors) {
    let mut buf: [u8; 20] = [0; 20];
    let enow = get_val(ENG_NOW, &mut buf);
    let eful = get_val(ENG_FUL, &mut buf);
    let used_pc = (enow / eful) * 100.0;

    cs.polyfmt_print("bat", "BAT");
    print!(" ");
    if used_pc > THRESHOLDS[0] {
        cs.polyfmt_print("good", &LIGHTNING_BOLT);
        return;
    }

    for i in 1..5 {
        if used_pc > THRESHOLDS[i] {
            print!("{:.2} {}", used_pc, UCHARS[i]);
            return;
        }
    }

    print!("{:.2} {}", used_pc, UCHARS[5]);
}
